src/backend/distributed/utils/meta_visibility.c

```diff
@@ -0,0 +1,505 @@
+	{
+		return false;
+	}
+
+	Form_pg_type typeForm = (Form_pg_type) GETSTRUCT(typeTuple);
+	Oid typeRelationId = typeForm->typrelid;
+
+	ReleaseSysCache(typeTuple);
+
+	/* check if type directly depends on citus */
+	if (IsCitusDependentObject(typeObjectAddress, NULL))
+	{
+		return true;
+	}
+
+	/* check if type's relation depends on citus, so type does indirectly */
+	if (OidIsValid(typeRelationId))
+	{
+		ObjectAddress typeRelationObjectAddress = {
+			RelationRelationId, typeRelationId, 0
+		};
+		return IsCitusDependentObject(typeRelationObjectAddress, NULL);
+	}
+
+	return false;
+}
+
+
+/*
+ * HideCitusDependentObjectsFromPgMetaTable adds a NOT is_citus_depended_object(oid, oid, smallint) expr
+ * to the security quals of meta class RTEs.
+ */
+bool
+HideCitusDependentObjectsFromPgMetaTable(Node *node, void *context)
+{
+	if (!CitusHasBeenLoaded())
+	{
+		return false;
+	}
+
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		MemoryContext queryContext = GetMemoryChunkContext(query);
+
+		/*
+		 * We process the whole rtable rather than visiting individual RangeTblEntry's
+		 * in the walker, since we need to know the varno to generate the right
+		 * filter.
+		 */
+		int varno = 0;
+		RangeTblEntry *rangeTableEntry = NULL;
+
+		foreach_ptr(rangeTableEntry, query->rtable)
+		{
+			varno++;
+
+			if (rangeTableEntry->rtekind == RTE_RELATION)
+			{
+				/* make sure the expression is in the right memory context */
+				MemoryContext originalContext = MemoryContextSwitchTo(queryContext);
+
+				Oid metaTableOid = InvalidOid;
+
+				/* add NOT is_citus_depended_object(oid, oid, oid) to the security quals of the RTE */
+				switch (rangeTableEntry->relid)
+				{
+					/* pg_class */
+					case RelationRelationId:
+
+					/* pg_proc */
+					case ProcedureRelationId:
+
+					/* pg_am */
+					case AccessMethodRelationId:
+
+					/* pg_type */
+					case TypeRelationId:
+
+					/* pg_enum */
+					case EnumRelationId:
+
+					/* pg_event_trigger */
+					case EventTriggerRelationId:
+
+					/* pg_trigger */
+					case TriggerRelationId:
+
+					/* pg_rewrite */
+					case RewriteRelationId:
+
+					/* pg_attrdef */
+					case AttrDefaultRelationId:
+
+					/* pg_constraint */
+					case ConstraintRelationId:
+
+					/* pg_ts_config */
+					case TSConfigRelationId:
+
+					/* pg_ts_template */
+					case TSTemplateRelationId:
+
+					/* pg_ts_dict */
+					case TSDictionaryRelationId:
+
+					/* pg_language */
+					case LanguageRelationId:
+
+					/* pg_namespace */
+					case NamespaceRelationId:
+
+					/* pg_sequence */
+					case SequenceRelationId:
+
+					/* pg_statistic */
+					case StatisticRelationId:
+
+					/* pg_attribute */
+					case AttributeRelationId:
+
+					/* pg_index */
+					case IndexRelationId:
+
+					/* pg_aggregate */
+					case AggregateRelationId:
+					{
+						metaTableOid = rangeTableEntry->relid;
+						break;
+					}
+
```

Comment:
do we really need this switch statement? Can't we simply rely on `metaTableOid = rangeTableEntry->relid;`

Reply:
We are only interested in pg meta relations specified by switch cases. We set the metaTableOid to invalidoid if it is not a pg meta relation.